pkg/controller/eventing/subscription/reconcile.go

```diff
@@ -0,0 +1,331 @@
+	// going to the call or if there's no call, directly to result.
+	from, err := r.resolveFromChannelable(subscription.Namespace, &subscription.Spec.From)
+	if err != nil {
+		glog.Warningf("Failed to resolve From %+v : %s", subscription.Spec.From, err)
+		return err
+	}
+	if from.Status.Subscribable == nil {
+		glog.Warningf("Failed to resolve the from %+v", subscription.Spec.From)
+		return fmt.Errorf("Failed to resolve the from %+v", subscription.Spec.From)
+	}
+
+	callDomain := ""
+	if subscription.Spec.Call != nil {
+		callDomain, err = r.resolveCall(subscription.Namespace, *subscription.Spec.Call)
+		if err != nil {
+			glog.Warningf("Failed to resolve Call %v : %v", *subscription.Spec.Call, err)
+		}
+		glog.Infof("Resolved call to: %q", callDomain)
+	}
+
+	resultDomain := ""
+	if subscription.Spec.Result != nil {
+		resultDomain, err = r.resolveResult(subscription.Namespace, *subscription.Spec.Result)
+		if err != nil {
+			glog.Warningf("Failed to resolve Result %v : %v", subscription.Spec.Result, err)
+		}
+		glog.Infof("Resolved result to: %q", resultDomain)
+	}
+
+	// Ok, now that we have the From and at least one of the Call/Result, let's greconcile
+	// the From with this information.
+	err = r.reconcileFromChannel(subscription.Namespace, from.Status.Subscribable.Channelable, callDomain, resultDomain, deletionTimestamp != nil)
+	if err != nil {
+		glog.Warningf("Failed to resolve from Channel : %s", err)
+		return err
+	}
+	return nil
+}
+
+func (r *reconciler) updateStatus(subscription *v1alpha1.Subscription) (*v1alpha1.Subscription, error) {
+	newSubscription := &v1alpha1.Subscription{}
+	err := r.client.Get(context.TODO(), client.ObjectKey{Namespace: subscription.Namespace, Name: subscription.Name}, newSubscription)
+
+	if err != nil {
+		return nil, err
+	}
+	newSubscription.Status = subscription.Status
+
+	// Until #38113 is merged, we must use Update instead of UpdateStatus to
+	// update the Status block of the Subscription resource. UpdateStatus will not
+	// allow changes to the Spec of the resource, which is ideal for ensuring
+	// nothing other than resource status has been updated.
+	if err = r.client.Update(context.TODO(), newSubscription); err != nil {
+		return nil, err
+	}
+	return newSubscription, nil
+}
+
+// resolveCall resolves the Spec.Call object. If it's an ObjectReference will resolve the object
+// and treat it as a Targetable.If it's TargetURI then it's used as is.
+// TODO: Once Service Routes, etc. support Targetable, use that.
+//
+func (r *reconciler) resolveCall(namespace string, callable v1alpha1.Callable) (string, error) {
+	if callable.TargetURI != nil && *callable.TargetURI != "" {
+		return *callable.TargetURI, nil
+	}
+
+	obj, err := r.fetchObjectReference(namespace, callable.Target)
+	if err != nil {
+		glog.Warningf("Feiled to fetch Callable target %+v: %s", callable.Target, err)
+		return "", err
+	}
+	t := duckv1alpha1.LegacyTarget{}
```

Comment:
This code seems to assume that every object reference is a LegacyTarget. Not necessary for this PR, but we should also support Targetable references, independent of Knative Serving's upgrade schedule.

Reply:
Yeah, agreed. Matt actually changed the serving stack to support Target.